### PR TITLE
[Package Renames 5] Transfer package popularity

### PR DESCRIPTION
**Pull request by loic-sharma**
This updates `auxiliary2azuresearch` to push download changes whenever a popularity transfer changes, or, whenever a download change affects a popularity transfer.

Previous changes: https://github.com/NuGet/NuGet.Services.Metadata/pull/765, https://github.com/NuGet/NuGet.Services.Metadata/pull/766, https://github.com/NuGet/NuGet.Services.Metadata/pull/767, and https://github.com/NuGet/NuGet.Services.Metadata/pull/768

Config changes: https://nuget.visualstudio.com/NuGetMicrosoft/_git/NuGetDeployment/pullrequest/1402?_a=overview

Part of https://github.com/nuget/nugetgallery/issues/7898

### CODE REVIEW

File: src/NuGet.Services.AzureSearch/DownloadTransferrer.cs [hunk 1]
```diff
@@ -64,6 +113,143 @@ public DownloadTransferrer(ILogger<DownloadTransferrer> logger)
...
+            // Next, calculate packages with incoming transfers. These packages receive downloads
+            // from one or more "outgoing" packages.
+            if (incomingTransfers.TryGetValue(packageId, out var incomingTransferIds))
+            {
+                var result = originalDownloads;
+
+                foreach (var incomingTransferId in incomingTransferIds)
+                {
+                    var incomingDownloads = downloads.GetDownloadCount(incomingTransferId);
+                    var incomingSplit = outgoingTransfers[incomingTransferId].Count;
```
Comment (joelverhagen) [inline src/NuGet.Services.AzureSearch/DownloadTransferrer.cs:241]: Is it possible for `incomingTransferId` to not be present in `outgoingTransfers`?

File: tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs [hunk 2]
```diff
@@ -28,6 +28,197 @@ public void ReturnsEmptyResult()
                 Assert.Empty(result);
             }
 
+            [Fact]
+            public void DoesNothingIfNoTransfers()
```
Comment (joelverhagen) [inline tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:32]: Consider adding an integration test with a stateful, in-memory blob service to simulate from transitions like:

- No rename to A -> B rename
- A -> B rename to A -> C rename
- A -> B rename to B -> A rename
- A -> B rename to A -> B, C -> D rename

I think those state transitions are the most likely source for bugs.

Comment (joelverhagen) [reply tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:32]: (this is in contrast to unit tests which may not cover the integration with the DAL)

File: tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs [hunk 3]
```diff
@@ -28,6 +28,197 @@ public void ReturnsEmptyResult()
...
+                Assert.Equal(200, result["C"]);
+            }
+
+            [Fact]
+            public void RejectsCyclicalPopularityTransfers()
+            {
+                PopularityTransfer = 1;
+
+                DownloadData.SetDownloadCount("A", "1.0.0", 100);
+                DownloadData.SetDownloadCount("B", "1.0.0", 100);
```
Comment (joelverhagen) [inline tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:190]: It's possible to "lose" a download, right? Like if the percentage against the total does no result in an exact number. We should have tests around that. It's okay to lose a download here or there but the behavior should be asserted IMO

**Comment by zhhyu** [comment]
>         {

Should we do null-check here considering the public method?

---
Refers to: src/NuGet.Services.AzureSearch/DownloadTransferrer.cs:69 in 032b459. [](commit_id = 032b45976f6f01bc7c53e05a7d65c98688e8f628, deletion_comment = False)

File: src/NuGet.Services.AzureSearch/DownloadTransferrer.cs [hunk 4]
```diff
@@ -23,9 +33,25 @@ public DownloadTransferrer(ILogger<DownloadTransferrer> logger)
             SortedDictionary<string, SortedSet<string>> outgoingTransfers,
             IReadOnlyDictionary<string, long> downloadOverrides)
         {
-            // TODO: Add download changes due to popularity transfers.
-            // See: https://github.com/NuGet/NuGetGallery/issues/7898
-            var downloadTransfers = new SortedDictionary<string, long>(StringComparer.OrdinalIgnoreCase);
+            Guard.Assert(
```
Comment (zhhyu) [inline src/NuGet.Services.AzureSearch/DownloadTransferrer.cs:36]: Should we have null-check here considering the public methods?

Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/DownloadTransferrer.cs:36]: We generally don't add null checks to our public methods.

File: src/NuGet.Services.AzureSearch/DownloadTransferrer.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGet.Services.AzureSearch/DownloadTransferrer.cs:241]: `incomingTransfers` is generated by inversing `outgoingTransfers`, so no, that should never happen.

File: tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs (hunk 3 above)
Comment (loic-sharma) [reply tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:190]: Yup, that's possible. I'll add test coverage 👍

Comment (loic-sharma) [reply tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:190]: Added

**Comment by zhhyu** [comment]
I am thinking that there is an edge case, but as I am not sure about the whole picture here, like how Azure Search actually works, so I just list it here:
If the original package is hard deleted, it seem that we will still keep it with the download counts 0, which will be fed to Azure Search, in my understanding. Is this expected?

**Comment by joelverhagen** [comment]
> If the original package is hard deleted, it seem that we will still keep it with the download counts 0, which will be fed to Azure Search, in my understanding. ("affectedPackages" will contain this deleted package Id.) Is this expected?

Great thing to be thinking about. Given renames are on the package registration level, deletes would only cascade to the rename record if we release the package registration, I think. Given this option is unchecked by default and only available if all versions are deleted, I think it would very rarely occur. This begs the question, if all versions are deleted, should we clear the rename?

**Comment by loic-sharma** [comment]
> If the original package is hard deleted, it seem that we will still keep it with the download counts 0....

Yup, you are correct that this will produce a transfer of `0`. This behavior is covered by the `DownloadTransferrerFacts.UnknownPackagesTransferZeroDownloads` test.

> ... the download counts 0, which will be fed to Azure Search

We generate a batch of "search index actions" to mutate the Azure Search index. 

The search pipeline has a "hidden" V3 resource called the version list, which is used by `catalog2azuresearch` and `auxiliary2azuresearch` to know which search documents should be updated. Using this resource, we detect and then skip updates to packages that do not exist. This behavior is covered by `SearchIndexActionBuilderFacts.UpdatesSearchDocumentsWithNoVersions`.

> This begs the question, if all versions are deleted, should we clear the rename?

The popularity transfer feature does little for packages with low download counts. I doubt we will need to clean "dangling" renames on "small" packages. Now say that we delete all versions of a popular package but keep its registration... This seems like a truly exceptional scenario, which is hard to anticipate. We may want to evaluate keeping/removing the popularity transfers on a case-by-case basis.

File: tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs (hunk 2 above)
Comment (loic-sharma) [reply tests/NuGet.Services.AzureSearch.Tests/DownloadTransferrerFacts.cs:32]: Added
